src/main/java/frc/robot/subsystems/DriveSubsystem.java

```diff
@@ -0,0 +1,185 @@
+
+            @Override
+            public void doCommand() {
+                frontLeftPositionController.resetPosition();
+                frontRightPositionController.resetPosition();
+                rearLeftPositionController.resetPosition();
+                rearRightPositionController.resetPosition();
+                logger.log("Clicks per rotation", rearRightEncoder.getCountsPerRevolution());
+            }
+        }.withTimeout(EntechCommandBase.DEFAULT_TIMEOUT_SECONDS);
+
+    }
+
+    @Override
+    public void initialize() {
+        frontLeftSpark = new CANSparkMax(RobotConstants.CAN.FRONT_LEFT_MOTOR, MotorType.kBrushless);
+        rearLeftSpark = new CANSparkMax(RobotConstants.CAN.REAR_LEFT_MOTOR, MotorType.kBrushless);
+        leftSpeedController = new SpeedControllerGroup(frontLeftSpark, rearLeftSpark);
+
+        frontRightSpark = new CANSparkMax(RobotConstants.CAN.FRONT_RIGHT_MOTOR, MotorType.kBrushless);
+        rearRightSpark = new CANSparkMax(RobotConstants.CAN.REAR_RIGHT_MOTOR, MotorType.kBrushless);
+        rightSpeedController = new SpeedControllerGroup(frontRightSpark, rearRightSpark);
+
+        frontLeftEncoder = frontLeftSpark.getEncoder();
+        frontRightEncoder = frontRightSpark.getEncoder();
+        rearLeftEncoder = rearLeftSpark.getEncoder();
+        rearRightEncoder = rearRightSpark.getEncoder();
+
+        robotDrive = new DifferentialDrive(leftSpeedController, rightSpeedController);
+
+        frontLeftPositionController = new SparkPositionController(frontLeftSpark, smartMotionSettings);
+        frontRightPositionController = new SparkPositionController(frontRightSpark, smartMotionSettings);
+        rearLeftPositionController = new SparkPositionController(rearLeftSpark, smartMotionSettings);
+        rearRightPositionController = new SparkPositionController(rearRightSpark, smartMotionSettings);
+        
+        posController = new SparkPositionControllerGroup(frontLeftPositionController, frontRightPositionController,
+                rearLeftPositionController, rearRightPositionController);
+        autoController = new PositionDriveController(frontLeftSpark, frontRightSpark, 
+                rearLeftSpark, rearRightSpark, smartMotionSettings, positionBuffer,
+                new EncoderInchesConverter(1/ RobotConstants.DIMENSIONS.MOTOR_REVOLUTIONS_PER_INCH));
+        reset();
+    }
+
+    public void setSpeedMode() {
+        speedSettings.configureSparkMax(frontLeftSpark);
+        speedSettings.configureSparkMax(frontRightSpark);
+        speedSettings.configureSparkMax(rearLeftSpark);
+        speedSettings.configureSparkMax(rearRightSpark);
+    }
+
+    public void setPositionMode() {
+        positionSettings.configureSparkMax(frontLeftSpark);
+        positionSettings.configureSparkMax(frontRightSpark);
+        positionSettings.configureSparkMax(rearLeftSpark);
+        positionSettings.configureSparkMax(rearRightSpark);
+    }
+
+    public EncoderValues getEncoderValues() {
```

Comment:
this is the right way to give robotposemanager a way to get the encoders. but I think we need to get the position from the positioncontroller objects, not the encoders directly, because the positioncontrollers know about the fact that one side may be inverted, and these methods do not.